src/Features/Core/Portable/ConvertAnonymousTypeToClass/AbstractConvertAnonymousTypeToClassCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,412 @@
+            }
+
+            context.RegisterRefactoring(new MyCodeAction(
+                c => ConvertToClassAsync(document, context.Span, c)));
+        }
+
+        private async Task<(TAnonymousObjectCreationExpressionSyntax, INamedTypeSymbol)> TryGetAnonymousObjectAsync(
+            Document document, TextSpan span, CancellationToken cancellationToken)
+        {
+            var position = span.Start;
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(position);
+
+            // Span actually has to be within the token (i.e. not in trivia around it).
+            if (!token.Span.IntersectsWith(position))
+            {
+                return default;
+            }
+
+            if (!span.IsEmpty && span != token.Span)
+            {
+                // if there is a selection, it has to be of the whole token.
+                return default;
+            }
+
+            var anonymousObject = token.Parent as TAnonymousObjectCreationExpressionSyntax;
+            if (anonymousObject == null)
+            {
+                return default;
+            }
+
+            // The position/selection must be of the 'new' token of the anonymous object.
+            if (anonymousObject.GetFirstToken() != token)
+            {
+                return default;
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var anonymousType = semanticModel.GetTypeInfo(anonymousObject, cancellationToken).Type as INamedTypeSymbol;
+
+            return (anonymousObject, anonymousType);
+        }
+
+        private async Task<Document> ConvertToClassAsync(
+            Document document, TextSpan span, CancellationToken cancellationToken)
+        {
+            var (anonymousObject, anonymousType) = await TryGetAnonymousObjectAsync(
+                document, span, cancellationToken).ConfigureAwait(false);
+
+            Debug.Assert(anonymousObject != null);
+            Debug.Assert(anonymousType != null);
+
+            var position = span.Start;
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var compilation = semanticModel.Compilation;
+
+            // Generate a unique name for the class we're creating.  We'll also add a rename
+            // annotation so the user can pick the right name for the type afterwards.
+            var className = NameGenerator.GenerateUniqueName(
+                "NewClass", n => semanticModel.LookupSymbols(position, name: n).IsEmpty);
+
+            // First, create the set of properties this class will have based on the properties
+            // the anonymous type has itself.
+            var (properties, propertyMap) = GenerateProperties(anonymousType);
```

Comment:
>propertyMap [](start = 29, length = 11)

I'd improve the comment to explain that the map is: anon property symbol -> new property name.

Reply:
Done.